src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFDataSourcesFactory.java

```diff
@@ -41,46 +41,97 @@ private XDDFDataSourcesFactory() {
     }
 
     public static XDDFCategoryDataSource fromDataSource(final CTAxDataSource categoryDS) {
-        return new XDDFCategoryDataSource() {
-            private CTStrData category = (CTStrData) categoryDS.getStrRef().getStrCache().copy();
-
-            @Override
-            public boolean isNumeric() {
-                return false;
-            }
-
-            @Override
-            public boolean isReference() {
-                return true;
-            }
-
-            @Override
-            public int getPointCount() {
-                return (int) category.getPtCount().getVal();
-            }
-
-            @Override
-            public String getPointAt(int index) {
-                return category.getPtArray(index).getV();
-            }
-
-            @Override
-            public String getDataRangeReference() {
-                return categoryDS.getStrRef().getF();
-            }
-
-            @Override
-            public int getColIndex() {
-                return 0;
-            }
-        };
+        if (categoryDS.getStrRef() == null) {
+            return new XDDFCategoryDataSource() {
```

Comment:
most of the code in the 2 anonymous classes is the same - would it be possible to create an inner class that has the shared code and to have the 2 anonymous classes extend from it?

Reply:
For sure!

Reply:
Like often with the OOXML code, it looks the same but has a different underlying class:
 `category` is either `CTNumData` or `CTStrData`... For the rest, I extracted the default implementation to the interface level.